### PR TITLE
partition: set `AddRecordOpt.isUpdate = true` when update partition records

### LINKED ISSUE #57588: Report unexpected duplicate when update partition tables during add column.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

``` golang
func TestUpdateMultiplePartitionTable(t *testing.T) {
	store := testkit.CreateMockStore(t)
	tk := testkit.NewTestKit(t, store)
	tk.MustExec("use test")
	tk.MustExec("create table t1 (c1 int, c2 int) partition by hash (c1) partitions 16")
	tk.MustExec("insert t1 values (1, 1), (2, 2)")
	tk.MustExec("create table t2 (c1 int, c2 int) partition by hash (c1) partitions 16")
	tk.MustExec("insert t2 values (1, 3), (2, 5)")
	tk2 := testkit.NewTestKit(t, store)
	tk2.MustExec("use test")

	testfailpoint.EnableCall(t, "github.com/pingcap/tidb/pkg/ddl/onJobUpdated", func(job *model.Job) {
		if job.SchemaState == model.StateWriteOnly {
			tk2.MustExec("update t1, t2 set t1.c1 = 8, t2.c2 = 10 where t1.c2 = t2.c1")
			tk2.MustQuery("select * from t1").Sort().Check(testkit.Rows("8 1", "8 2"))
			tk2.MustQuery("select * from t2").Sort().Check(testkit.Rows("1 10", "2 10"))
		}
	})

	tk.MustExec("alter table t1 add column c3 bigint default 9")

	tk.MustQuery("select * from t1").Sort().Check(testkit.Rows("8 1 9", "8 2 9"))
}
```

### 2. What did you expect to see? (Required)

test pass

### 3. What did you see instead (Required)

test failed

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #57588

Problem Summary: 

Didn't set `isUpdate = true` before, so it will use the last value in datum as `_tidb_rowid` which caused primary key duplicate.

If we don't do this during add columns, `len(r) == len(cols)` always equal to true, so it's safe.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/table/table.go [hunk 1]
```diff
@@ -202,7 +202,7 @@ func (opt *UpdateRecordOpt) SkipWriteUntouchedIndices() bool {
 
 // GetAddRecordOpt creates a AddRecordOpt.
 func (opt *UpdateRecordOpt) GetAddRecordOpt() *AddRecordOpt {
-	return &AddRecordOpt{commonMutateOpt: opt.commonMutateOpt}
+	return &AddRecordOpt{commonMutateOpt: opt.commonMutateOpt, isUpdate: true}
```
Comment (mjonss) [inline pkg/table/table.go:205]: Is this only needed for partitioned tables, or also for non-partitioned tables, where the PK/_tidb_rowid is updated or replaced (like REPLACE INTO)?

Comment (Defined2014) [reply pkg/table/table.go:205]: non-partitioned table handles it. https://github.com/pingcap/tidb/blob/9ef5b59f6dc0ddac6b1151d4b62368cfca2457ec/pkg/executor/write.go#L184

**Review by mjonss — APPROVED** [review]
LGTM

**Review by tangenta — APPROVED** [review]

**Pull request merged**